File: darts-suite/fortran/darts-mpi.f
! Compute pi using six basic MPI subroutines
        program darts
          use lcgenerator
          implicit none
          include 'mpif.h'
          integer :: num_trials = 1000000, i = 0, Ncirc = 0
          real :: pi = 0.0, x = 0.0, y = 0.0, r = 1.0
          real :: r2 = 0.0

          integer :: rank, np, manager = 0
          integer :: mpistatus, mpierr, j
          integer :: my_trials, temp

          call MPI_Init(mpierr)
          call MPI_Comm_size(MPI_COMM_WORLD, np, mpierr)
          call MPI_Comm_rank(MPI_COMM_WORLD, rank, mpierr)
          r2 = r*r
          my_trials = num_trials/np
          if (mod(num_trials, np) .gt. rank) then
           my_trials = my_trials+1
          end if
          call seed(rank)

          do i = 1, my_trials
            x = lcgrandom()
            y = lcgrandom()
            if ((x*x + y*y) .le. r2) then
              Ncirc = Ncirc+1
            end if
          end do

          if (rank .eq. manager) then
            do j = 1, np-1
              call MPI_Recv(temp, 1, MPI_INTEGER, j, j, 
     &          MPI_COMM_WORLD, mpistatus, mpierr)
              Ncirc = Ncirc + temp
            end do
            pi = 4.0*((1.0*Ncirc)/(1.0*num_trials))
            print*, '     '
            print*, '     Computing pi using six basic MPI functions:'
            print*, '     For ', num_trials, ' trials, pi = ', pi
            print*, '     '
          else
            call MPI_Send(Ncirc, 1, MPI_INTEGER, manager, rank, 
     &        MPI_COMM_WORLD, mpierr) 
          end if
          call MPI_Finalize(mpierr)
        end

File: darts-suite/c/darts-mpi.c
/* Compute pi using the six basic MPI functions */
#include "lcgenerator.h"
#include <mpi.h>
#include <stdio.h>

int main(int argc, char **argv) {

    long num_trials = 1000000;
    long local_Ncirc = 0;
    long total_Ncirc = 0;
    long received_Ncirc;
    double pi, x, y;
    double r = 1.0;
    double r2 = r*r;
    long base_seed = 12345;
    
    int rank, size, manager = 0, source;
    MPI_Status status;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    long my_trials = num_trials / size;

    for (long i = 0; i < my_trials; i++) {
        long thread_seed = base_seed + (rank + (i * size)) * 2L;

        x = lcgrandom_r(&thread_seed);
        y = lcgrandom_r(&thread_seed);

        if ((x*x + y*y) <= r2) {
            local_Ncirc++;
        }
    }

    if (rank == manager) {
        total_Ncirc = local_Ncirc;

        for (source = 1; source < size; source++) {
            MPI_Recv(&received_Ncirc, 1, MPI_LONG, source, 0, MPI_COMM_WORLD, &status);
            total_Ncirc += received_Ncirc;
        }

        pi = 4.0 * ((double)total_Ncirc) / ((double)num_trials);
        printf("\n \t Computing pi using six basic MPI functions: \n");
        printf("\t For %ld trials, pi = %f\n", num_trials, pi);
        printf("\n");
    } 
    else {
        MPI_Send(&local_Ncirc, 1, MPI_LONG, manager, 0, MPI_COMM_WORLD);
    }

    MPI_Finalize();
    return 0;
}
